### PR TITLE
Wallabag plugin

### PR DESCRIPTION
**Pull request by y-muller**
A plugin to download articles from a Wallabag server, an open source read-it-later service (https://wallabag.org).

Issues:
- (solved) The settings menu changes the menu item text for the download directory and the tag filter. It makes it easy to see the current configuration. But I have not found how to refresh the value after a change. I could remove the feature but I'd rather have a hint on how to fix it.

- To skip files that already exist locally, I want to check the dates of the local and remote versions. The only date parser is in the newsdownloader plugin. I have added an ugly hack to use it if available. Thoughts?

### CODE REVIEW

**Comment by poire-z** [comment]
About your 1st issue _I have not found how to refresh the value after a change_ :

the callback gets provided the `touchmenu_instance` as its first argument, so you can do something with it inside the callback depending on what happened (most often: close it, or update/refresh it). See some examples in #4189, eg: https://github.com/koreader/koreader/pull/4189/files#diff-01418c46265dd5757e5be1103a3a224e).
In your case, you want to pass it to `Wallabag:setDownloadDirectory(touchmenu_instance)`, which can call just after `self:saveSettings()` : `touchmenu_instance:updateItems()`

**Comment by y-muller** [comment]
Can I run the CI checks on my own branch?

**Comment by Frenzie** [comment]
You can enable CircleCI on your own repo, yes. Let me know if you need a pointer!

File: plugins/wallabag.koplugin/main.lua [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+local ltn12 = require('ltn12')
+local http = require('socket.http')
+local https = require('ssl.https')
+local socket = require('socket')
+local url = require('socket.url')
+local util = require("util")
+local _ = require("gettext")
+local T = FFIUtil.template
+local Screen = require("device").screen
+local DataStorage = require("datastorage")
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Please note the order in our coding standards.

File: plugins/wallabag.koplugin/main.lua [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+                            end
+                            return _("Filter articles by tag") .. " (" .. filter .. ")"
+                        end,
+                        keep_menu_open = true,
+                        callback = function(touchmenu_instance)
+                            self:setFilterTag(touchmenu_instance)
+                        end,
+                    },
+                    {
+                        text = _("Remotely delete Finished articles"),
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Capitalization.

File: plugins/wallabag.koplugin/main.lua [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+                        checked_func = function() return self.is_delete_finished end,
+                        callback = function()
+                            self.is_delete_finished = not self.is_delete_finished
+                            self:saveSettings()
+                        end,
+                    },
+                    {
+                        text = _("Remotely delete 100% read articles"),
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Or just finished?

File: plugins/wallabag.koplugin/main.lua [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+                            self.is_sync_remote_delete = not self.is_sync_remote_delete
+                            self:saveSettings()
+                        end,
+                    },
+                    {
+                        text = _("Help"),
+                        keep_menu_open = true,
+                        callback = function()
+                            UIManager:show(InfoMessage:new{
+                                text = _('Download directory: use a directory that is exclusively used by the Wallabag plugin. ' ..
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Please use a multiline comment. Our translation string extraction regex will fail on this.

File: plugins/wallabag.koplugin/main.lua [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+                        end,
+                    }
+                }
+            },
+            {
+                text = _("Info"),
+                keep_menu_open = true,
+                callback = function()
+                    UIManager:show(InfoMessage:new{
+                        text = T(_('Wallabag is an open source Read-it-later service. This plugin synchronises with a Wallabag server.\n\n' ..
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: See above

File: plugins/wallabag.koplugin/main.lua [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+    if not self.is_dateparser_checked then
+        local res
+        res, self.dateparser = pcall(require, "lib.dateparser")
+        if res then self.is_dateparser_available = true end
+        self.is_dateparser_checked = true
+    end
+
+    -- Check if the configuration is complete
+    local function isempty(s)
+        return s == nil or s == ''
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Nitpick, but could you use double quotes?

File: plugins/wallabag.koplugin/main.lua [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+                    })
+                end,
+            },
+        },
+    }
+end
+
+function Wallabag:getBearerToken()
+    -- workaround for dateparser, only once
+    -- the parser is in newsdownloader.koplugin, check if it is available
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Should it be moved?

File: plugins/wallabag.koplugin/main.lua [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+                }
+            }
+        },
+    }
+    UIManager:show(self.tag_dialog)
+    self.tag_dialog:onShowKeyboard()
+end
+
+function Wallabag:editServerSettings()
+    local text_info = "Enter the details of your Wallabag server and account.\n"..
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Multiline

File: plugins/wallabag.koplugin/main.lua (hunk 4 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: Is a max of 79 chars OK or should I split them further?

File: plugins/wallabag.koplugin/main.lua (hunk 1 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: Done.

File: plugins/wallabag.koplugin/main.lua (hunk 7 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: I moved it to init().

File: plugins/wallabag.koplugin/main.lua (hunk 3 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: I want to make a distinction between articles that have been read to the end (100% progress) and articles that have been marked as finished/complete in the status screen.

That makes it possible to delete only the articles where the user took the extra step of setting the status, for example.

File: plugins/wallabag.koplugin/main.lua (hunk 8 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: Shortened strings a bit.

Comment (Frenzie) [reply plugins/wallabag.koplugin/main.lua]: It's not about length but concat won't work right.

Use either just a single line with \n or [[multiline]]

Comment (Frenzie) [reply plugins/wallabag.koplugin/main.lua]: It's not really a defect in the regex imo (or I would've changed it) because [[]] is much nicer anyway.

Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: I get it now. I have changed it to multiline.

File: plugins/wallabag.koplugin/main.lua [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+            {
+                text = _("Delete finished articles remotely"),
+                callback = function()
+                    if not NetworkMgr:isOnline() then
+                        NetworkMgr:promptWifiOn()
+                        return
+                    end
+                    local num_deleted = self:processLocalFiles( "manual" )
+                    UIManager:show(InfoMessage:new{
+                        text = T(_('Articles processed.\nDeleted: %1'), num_deleted)
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Could you also make these double quotes for consistency?

File: plugins/wallabag.koplugin/main.lua [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+                    },
+                    {
+                        text_func = function()
+                            local path
+                            if not self.directory or self.directory == "" then
+                                path = "Not set"
+                            else
+                                path = filemanagerutil.abbreviate(self.directory)
+                            end
+                            return _("Set download directory") .. " (" .. path .. ")"
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Even for stuff like this it's generally preferable to use T(). The thing is, you don't really know if in some localization might like to make it `(path) Set download directory` or `Download directory (path) set` or many other possibilities. :-)

File: plugins/wallabag.koplugin/main.lua [hunk 11]
```diff
@@ -0,0 +1,695 @@
...
+                        end,
+                    }
+                }
+            },
+            {
+                text = _("Info"),
+                keep_menu_open = true,
+                callback = function()
+                    UIManager:show(InfoMessage:new{
+                        text = T(_([[Wallabag is an open source Read-it-later service. This plugin synchronises with a Wallabag server.
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: read-it-later should be lowercase

File: plugins/wallabag.koplugin/main.lua [hunk 12]
```diff
@@ -0,0 +1,695 @@
...
+function Wallabag:getBearerToken()
+
+    -- Check if the configuration is complete
+    local function isempty(s)
+        return s == nil or s == ""
+    end
+
+    if isempty(self.server_url) or isempty(self.username) or isempty(self.password)  or isempty(self.client_id) or isempty(self.client_secret) or isempty(self.directory) then
+        UIManager:show(InfoMessage:new{
+            text = _('Please configure the server and local settings.')
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: quotes

File: plugins/wallabag.koplugin/main.lua [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+        })
+        return false
+    end
+
+    -- Check if the download directory is valid
+    local dir_mode = lfs.attributes(self.directory, "mode")
+    logger.dbg("mode:", dir_mode)
+    if dir_mode ~= "directory" then
+         UIManager:show(InfoMessage:new{
+            text = _('The download directory is not valid.\nPlease configure it in the settings.')
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: quotes

File: plugins/wallabag.koplugin/main.lua [hunk 14]
```diff
@@ -0,0 +1,695 @@
...
+        })
+        return false
+    end
+    if string.sub( self.directory, -1 ) ~= '/' then
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: more quotes ;-)

File: plugins/wallabag.koplugin/main.lua [hunk 15]
```diff
@@ -0,0 +1,695 @@
...
+    end
+
+    local now = os.time()
+    if self.token_expiry - now > 300 then
+        -- token still valid for a while, no need to renew
+        return true
+    end
+
+    local login_url = "/oauth/v2/token"
+    local body = "{ \"grant_type\": \"password\", \"client_id\": \"" .. self.client_id .. "\", \"client_secret\": \"" .. self.client_secret .. "\", \"username\": \"" .. self.username .. "\", \"password\": \"" .. self.password .. "\"}"
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Yet perhaps here there's a good reason to deviate and use single quotes? ;-)

Comment (Frenzie) [reply plugins/wallabag.koplugin/main.lua]: PS string.format could potentially be more legible.

File: plugins/wallabag.koplugin/main.lua [hunk 16]
```diff
@@ -0,0 +1,695 @@
...
+    local headers = {
+        ["Content-type"] = "application/json",
+        ["Accept"] = "application/json, */*",
+        ["Content-Length"] = tostring(#body),
+        }
+    local result = self:callAPI( 'POST', login_url, headers, body, "" )
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: I have to wonder actually, is there a specific rationale behind where you used single and where double quotes?

File: plugins/wallabag.koplugin/main.lua [hunk 17]
```diff
@@ -0,0 +1,695 @@
...
+    else
+        UIManager:show(InfoMessage:new{
+            text = _("Could not login to Wallabag server."), })
+        return false
+    end
+end
+
+function Wallabag:getArticleList()
+    local articles_url = "/api/entries.json?archive=0&tags=" .. self.filter_tag
+    return self:callAPI( 'GET', articles_url, nil, "", "" )
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: quotes

File: plugins/wallabag.koplugin/main.lua [hunk 18]
```diff
@@ -0,0 +1,695 @@
...
+
+-- method: (mandatory) GET, POST, DELETE, PATCH, etc...
+-- apiurl: (mandatory) excluding the server path
+-- headers: defaults to auth if given nil value, provide all headers necessary if in use
+-- body: empty string if not needed
+-- filepath: downloads the file if provided, returns JSON otherwise
+function Wallabag:callAPI( method, apiurl, headers, body, filepath )
+    local request, sink = {}, {}
+    local parsed = url.parse(apiurl)
+    request['url'] = self.server_url .. apiurl
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: more quotes

File: plugins/wallabag.koplugin/main.lua [hunk 19]
```diff
@@ -0,0 +1,695 @@
...
+    end
+    request['headers'] = headers
+    if body ~= "" then
+        request['source'] = ltn12.source.string(body)
+    end
+    logger.dbg("URL     ", self.server_url .. apiurl)
+    logger.dbg("method  ", method)
+
+    http.TIMEOUT, https.TIMEOUT = 10, 10
+    local httpRequest = parsed.scheme == 'http' and http.request or https.request
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: quotes

File: plugins/wallabag.koplugin/main.lua [hunk 20]
```diff
@@ -0,0 +1,695 @@
...
+        end -- for entry
+    end -- flag checks
+    return num_deleted
+end
+
+function Wallabag:deleteArticle( path )
+    logger.dbg("deleting article ", path )
+    local id = self:getArticleID( path )
+    if id then
+        self:callAPI( 'DELETE', "/api/entries/" .. id .. ".json", nil, "", "" )
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: quotes

File: plugins/wallabag.koplugin/main.lua [hunk 21]
```diff
@@ -0,0 +1,695 @@
...
+    }
+    UIManager:show(self.tag_dialog)
+    self.tag_dialog:onShowKeyboard()
+end
+
+function Wallabag:editServerSettings()
+    local text_info = [[Enter the details of your Wallabag server and account.
+
+Client ID and client secret are long strings so you might prefer to save the empty settings and edit the config file directly:
+.adds/koreader/settings/wallabag.lua
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: This is Kobo-specific (and even then mainly for KSM), plus I suspect you typod settings? You should populate this with `DataStorage:getSettingsDir()`.

https://github.com/koreader/koreader/blob/fa0117bb2a02d0984bded06144c93218a85b21cb/datastorage.lua#L36

File: plugins/wallabag.koplugin/main.lua [hunk 22]
```diff
@@ -0,0 +1,695 @@
...
+                    },
+                    {
+                        text_func = function()
+                            local filter
+                            if not self.filter_tag or self.filter_tag == "" then
+                                filter = "All articles"
+                            else
+                                filter = self.filter_tag
+                            end
+                            return _("Filter articles by tag") .. " (" .. filter .. ")"
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua]: Copying my other comment.

> Even for stuff like this it's generally preferable to use T().

https://github.com/koreader/koreader/pull/4271/files#r225139189

File: plugins/wallabag.koplugin/main.lua (hunk 15 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: Good shout. A string.format and a [[multiline]] made it a lot neater.

File: plugins/wallabag.koplugin/main.lua (hunk 16 above)
Comment (y-muller) [reply plugins/wallabag.koplugin/main.lua]: I'm afraid there was no specific logic in my use of quotes.
I starting by grabbing a few bits of interesting code in other plugins and after a while the inconsistencies spread a bit.

All changed to double quotes and a few multilines.

**Comment by Frenzie** [comment]
Alright, I'll let it sit a couple more days so feel free to ping me then if I haven't looked on GH yet.

@poire-z or just merge it if you think it's ready. ;-)

**Comment by poire-z** [comment]
Looking fine and clean, so merging.

**Comment by poire-z** [comment]
There might be a little something if one has not yet configured this plugin.
Not yet tried on my kobo, but on the emulator:
```
10/17/18-22:30:49 INFO  FM loaded plugin terminal at plugins/terminal.koplugin
10/17/18-22:30:49 INFO  FM loaded plugin texteditor at plugins/texteditor.koplugin
10/17/18-22:30:49 ERROR Failed to initialize wallabag plugin:  plugins/wallabag.koplugin/main.lua:44: attempt to index field 'wallabag' (a nil value)
```
The menu is still registered and usable thus.

https://github.com/koreader/koreader/blob/a3a17dbbeb7bcd94702c2f12acf1e09809340b50/plugins/wallabag.koplugin/main.lua#L42-L45

Looks like when there is no settings yet, there is no `self.wb_settings.data.wallabag` table.
Might just need to be set to `...wallabag={}` in `Wallabag:readSettings()` if it's not there.

**Comment by bobberb** [comment]
The wallabag plugin provides ALMOST 1:1 parity to the pocket integration on the Kobo. This is amazing. Currently, the functionality is focused on server-side deletion post-read. 

To accommodate maybe more use cases, maybe, there should be an option that only favorited/unread articled be allowed. Triggers activate when the un-stared article is finished to make the "mark as read" api call. On next sync the article will be removed from the device but kept on the server. At that point, the device could also accept permanently retained articles via a "device" tag.

**Comment by y-muller** [comment]
Thanks @bobberb. I have a couple of bugs to fix on the plugin first. Then I'll have a look at your suggestion.

**Pull request merged**